File: src/main/java/org/iuc/mobileteam/iucnewsappbackend/controllers/AnnouncementsController.java
package org.iuc.mobileteam.iucnewsappbackend.controllers;

import org.iuc.mobileteam.iucnewsappbackend.dtos.AnnouncementsDto;
import org.iuc.mobileteam.iucnewsappbackend.services.AnnouncementsService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/announcements")
public class AnnouncementsController {
    private final static Logger LOGGER = LoggerFactory.getLogger(AnnouncementsController.class);
    private final AnnouncementsService announcementsService;

    public AnnouncementsController(AnnouncementsService announcementsService) {
        this.announcementsService = announcementsService;
    }

    @PostMapping()
    @ResponseStatus(HttpStatus.CREATED)
    public List<AnnouncementsDto> postAnnouncements(@RequestBody List<AnnouncementsDto> announcements) {
        LOGGER.info("Received POST request to create announcements");
        return announcementsService.saveAll(announcements);
    }

    @GetMapping()
    public List<AnnouncementsDto> getAnnouncements() {
        LOGGER.info("Received GET request to fetch announcements");
        return announcementsService.getAll();
    }

    @GetMapping("/{id}")
    public AnnouncementsDto getAnnouncementById(@PathVariable String id) {
        LOGGER.info("Received GET request to fetch announcement with id: {}", id);
        return announcementsService.getById(id);
    }
    @DeleteMapping()
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public long deleteAnnouncements() {
        LOGGER.info("Received DELETE request to delete all announcements");
        return announcementsService.deleteAll();
    }

    @ExceptionHandler(RuntimeException.class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public final Exception handleAllExceptions(RuntimeException e) {
        LOGGER.error("Internal server error.", e);
        return e;
    }
}

File: src/main/kotlin/com/iucmobileteam/iucnewsbackend/controller/AnnouncementsController.kt
package com.iucmobileteam.iucnewsbackend.controller

import com.iucmobileteam.iucnewsbackend.dtos.AnnouncementsDto
import com.iucmobileteam.iucnewsbackend.entity.AnnouncementsEntity
import com.iucmobileteam.iucnewsbackend.service.AnnouncementsService
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus

@RestController
@RequestMapping("announcements")
class AnnouncementsController (private val announcementsService: AnnouncementsService) {
    private val logger: Logger = LoggerFactory.getLogger(AnnouncementsController::class.java)

    @GetMapping
    fun getAnnouncements(): MutableList<AnnouncementsEntity> = announcementsService.getAnnouncements()

    @GetMapping("{id}")
    fun getAnnouncementById(@PathVariable id: Long): Optional<AnnouncementsEntity> = announcementsService.getAnnouncementById(id)

    @PostMapping()
    @ResponseStatus(HttpStatus.CREATED)
    fun createAnnouncement(@RequestBody announcement: AnnouncementsDto): AnnouncementsEntity = announcementsService.createAnnouncement(announcement.toAnnouncementsEntity())

    @PostMapping("bulk")
    @ResponseStatus(HttpStatus.CREATED)
    fun updateAnnouncements(@RequestBody announcements: List<AnnouncementsDto>): MutableList<AnnouncementsEntity> = announcementsService.updateAnnouncements(announcements.map { it.toAnnouncementsEntity() })

    @DeleteMapping()
    fun deleteAllAnnouncements() = announcementsService.deleteAllAnnouncements()

    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    final fun handleAllExceptions(e: Exception): Exception {
        logger.error("An error occurred: ${e.message}")
        return e
    }
}